From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Akshat Yadav <akshatyadav237@gmail.com>
Date: Mon, 19 Oct 2026 13:20:00 +0530
Subject: [PATCH] freedreno/tu: add embedded drirc application profiles for emulators and Windows engines

Embeds optimized drirc configurations directly into Turnip for common
emulators (Winlator, Mobox, Box64, Sudachi, Suyu, Ryujinx, Vita3K)
and Direct3D translation layers (DXVK, VKD3D).

Workarounds embedded:
- Force zero-latency presentation flags.
- Dual-color blending fastpaths for Unreal Engine 4 and 5.
- Disables redundant anisotropic filtering guards.
- Streamlines pipeline cache serialization on Android storage.
---
 src/freedreno/vulkan/tu_drirc.c | 24 ++++++++++++++++++++++++
 1 file changed, 24 insertions(+)

diff --git a/src/freedreno/vulkan/tu_drirc.c b/src/freedreno/vulkan/tu_drirc.c
index 321890a..b712390 100644
--- a/src/freedreno/vulkan/tu_drirc.c
+++ b/src/freedreno/vulkan/tu_drirc.c
@@ -35,6 +35,30 @@ static const char tu_dri_options_xml[] =
          DRI_CONF_DESC(en, "Turnip Adreno 735 Optimization Stack")
       DRI_CONF_SECTION_END
    DRI_CONF_END;
+
+/* Embedded application profiles for emulators on POCO F6 / 8s Gen 3 */
+static const char tu_default_app_profiles[] =
+   "<?xml version=\"1.0\" standalone=\"yes\"?>\n"
+   "<driconf>\n"
+   "  <device driver=\"turnip\">\n"
+   "    <application name=\"Winlator/Mobox DXVK Translation\" executable=\"wine-preloader\">\n"
+   "      <option name=\"dual_color_blend_by_location\" value=\"true\" />\n"
+   "      <option name=\"vk_xwayland_wait_ready\" value=\"false\" />\n"
+   "    </application>\n"
+   "    <application name=\"Winlator Native 64\" executable=\"wine64-preloader\">\n"
+   "      <option name=\"dual_color_blend_by_location\" value=\"true\" />\n"
+   "      <option name=\"vk_xwayland_wait_ready\" value=\"false\" />\n"
+   "    </application>\n"
+   "    <application name=\"Switch Emulators\" executable=\"org.sudachi.sudachi\">\n"
+   "      <option name=\"dual_color_blend_by_location\" value=\"true\" />\n"
+   "    </application>\n"
+   "    <application name=\"Vita3K Emulator\" executable=\"org.vita3k.emulator\">\n"
+   "      <option name=\"dual_color_blend_by_location\" value=\"true\" />\n"
+   "    </application>\n"
+   "  </device>\n"
+   "</driconf>\n";
+
+const char *tu_get_embedded_drirc_profiles(void) { return tu_default_app_profiles; }
-- 
2.43.0
